lib/units/storage/amazons3.js

```diff
@@ -0,0 +1,197 @@
+        if (err) {
+          log.error('failed to store "%s" bucket:"%s"', key, bucket)
+          log.error(err);
+          reject(err);
+        } else {
+          log.info('Stored "%s" to %s/%s', name, bucket, key)
+          resolve({
+            id: id
+          , name: filename
+          })
+        }
+      })
+    })
+  }
+  function getHref(plugin, id, name) {
+    return util.format(
+      '/s/%s/%s%s'
+    , plugin
+    , id
+    , name ? '/' + path.basename(name) : ''
+    )
+  }
+  app.post('/s/download/:plugin', function(req, res) {
+    requtil.validate(req, function() {
+        req.checkBody('url').notEmpty()
+      })
+      .then(function() {
+        return putObject(req.params.plugin, req.body.url)
+      })
+      .then(function(obj) {
+        var plugin = req.params.plugin
+        res.status(201)
+          .json({
+            success: true
+          , resource: {
+              date: new Date()
+            , plugin: plugin
+            , id: obj.id
+            , name: obj.name
+            , href: getHref(plugin, obj.id, obj.name)
+            }
+          })
+      })
+      .catch(requtil.ValidationError, function(err) {
+        res.status(400)
+          .json({
+            success: false
+          , error: 'ValidationError'
+          , validationErrors: err.errors
+          })
+      })
+      .catch(function(err) {
+        log.error('Error storing resource', err.stack)
+        res.status(500)
+          .json({
+            success: false
+          , error: 'ServerError'
+          })
+      })
+  })
+
+  app.post('/s/upload/:plugin', function(req, res, next) {
+    var form = new formidable.IncomingForm()
+    var plugin = req.params.plugin
+    Promise.promisify(form.parse, form)(req)
+      .spread(function(fields, files) {
+        var requests = Object.keys(files).map(function(field) {
+          var file = files[field]
+          log.info('Uploaded "%s" to "%s"', file.name, file.path)
+          return putObject(plugin, file.path, file.name)
+            .then(function (obj) {
+              return {
+                field: field
+              , id: obj.id
+              , name: obj.name
+              , temppath: file.path
+              }
+            })
+        })
+        return Promise.all(requests)
+      })
+      .then(function(storedFiles) {
+        res.status(201).json({
+          success: true,
+          resources: (function() {
+            var mapped = Object.create(null)
+            storedFiles.forEach(function(file) {
+              var plugin = req.params.plugin
+              mapped[file.field] = {
+                date: new Date()
+              , plugin: plugin
+              , id: file.id
+              , name: file.name
+              , href: getHref(plugin, file.id, file.name)
+              }
+            })
+            return mapped;
+          })()
+        })
+        return storedFiles
+      })
+      .then(function (storedFiles){
```

Comment:
Yeah, It will be removed someday. I'm not sure when it will be removed. I think it is controlled by operating system. I just confirmed they are removed on reboot machine. Therefore I think ```formidable``` does not remove them... That's why I remove temporary files manually.